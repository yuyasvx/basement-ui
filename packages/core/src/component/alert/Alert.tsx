import { FC, PropsWithChildren, ReactNode, useMemo } from 'react';
import clsx from 'clsx';
import { RootStyle } from '../../domain/StyleClass';
import { BaseComponentProps, getBaseComponentProps } from '../../base/BaseComponent';
import { Case } from '../../util/Case';
import { MouseEvents } from '../../domain/EventProps';
import { getMouseEventHandler } from '../../util/Handler';
import { AlertContent, AlertContentDetailedProps } from './AlertContent';

const NAME = 'bm-c-alert';
const INNER_NAME = `${NAME}__inner`;
const ICON_NAME = `${NAME}__icon`;
const DETAIL_NAME = `${NAME}__detail`;

export const AlertContentLayout = {
  HORIZONTAL: 'horizontal',
  VERTICAL: 'vertical'
} as const;

interface AlertDetailedProps {
  layout?: Case<typeof AlertContentLayout>;
  centered?: boolean;
  icon?: ReactNode;
  iconPosition?: 'left' | 'right';
  // footerType?: 'default' | 'hide' | 'custom';
}

export type AlertProps = PropsWithChildren<
  AlertDetailedProps & AlertContentDetailedProps & BaseComponentProps & MouseEvents<HTMLDivElement>
>;

export const useAlertHook = (props: AlertProps) => {
  const layoutClassName = useMemo(() => {
    const l = props.layout ?? AlertContentLayout.HORIZONTAL;
    return `-${l}`;
  }, [props.layout]);
  const centeredClassName = useMemo(() => {
    const n = props.centered ?? false;
    return n ? `-centered` : '';
  }, [props.centered]);
  const iconPositionClassName = useMemo(() => {
    const l = props.iconPosition ?? 'left';
    return `-icon-${l}`;
  }, [props.iconPosition]);
  const alertContentClassName = useMemo(
    () => (props.layout === AlertContentLayout.VERTICAL && props.centered ? '-centered' : ''),
    [props.centered, props.layout]
  );

  const bp = getBaseComponentProps(props);
  const me = getMouseEventHandler(props);

  const classNames = clsx(
    NAME,
    RootStyle.BASE,
    RootStyle.CONTENT_BASE,
    props.className,
    layoutClassName,
    centeredClassName,
    iconPositionClassName
  );

  return {
    mainProps: {
      className: classNames,
      ...bp,
      ...me
    },
    innerProps: {
      className: INNER_NAME
    },
    iconProps: {
      className: ICON_NAME
    },
    contentProps: {
      title: props.title,
      footer: props.footer,
      className: useMemo(() => clsx(DETAIL_NAME, alertContentClassName), [alertContentClassName])
    }
  };
};
export const Alert: FC<AlertProps> = props => {
  const { mainProps, innerProps, contentProps, iconProps } = useAlertHook(props);

  return (
    <div {...mainProps}>
      <div {...innerProps}>
        {props.icon && <div {...iconProps}>{props.icon}</div>}
        <AlertContent {...contentProps}>{props.children}</AlertContent>
      </div>
    </div>
  );
};
