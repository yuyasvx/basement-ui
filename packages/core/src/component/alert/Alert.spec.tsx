import { cleanup, render, renderHook } from '@testing-library/react';
import { Alert, AlertContentLayout, useAlertComponent } from './Alert';

describe('Alert Component Snapshot', () => {
  afterEach(() => {
    cleanup();
  });

  test('Propsなし', () => {
    const { asFragment } = render(<Alert />);
    expect(asFragment()).toMatchSnapshot();
  });

  test('Propsあり_レイアウト縦_中央揃えあり', () => {
    const { asFragment } = render(
      <Alert
        layout={AlertContentLayout.VERTICAL}
        centered
        icon={<div>icon</div>}
        iconPosition="right"
        title={<strong>title</strong>}
        footer={<div>footer</div>}
        footerPosition="right"
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test('Propsあり_レイアウト縦_中央揃えなし', () => {
    const { asFragment } = render(
      <Alert
        layout={AlertContentLayout.VERTICAL}
        centered={false}
        icon={<div>icon</div>}
        iconPosition="right"
        title={<strong>title</strong>}
        footer={<div>footer</div>}
        footerPosition="right"
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test('Propsあり_レイアウト横', () => {
    const { asFragment } = render(
      <Alert
        layout={AlertContentLayout.HORIZONTAL}
        centered
        icon={<div>icon</div>}
        iconPosition="right"
        title={<strong>title</strong>}
        footer={<div>footer</div>}
        footerPosition="right"
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});

describe('Alert Component Hook', () => {
  it('', () => {
    const { result } = renderHook(() => useAlertComponent({}));
    expect(result.current.mainProps.className).toBe(`bm-c-alert bm-base bm-content-base -horizontal -icon-left`);
  });
});
