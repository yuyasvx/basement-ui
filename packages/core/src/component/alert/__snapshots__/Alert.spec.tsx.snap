// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Alert Component Snapshot Propsあり_レイアウト横 1`] = `
<DocumentFragment>
  <div
    class="bm-c-alert bm-base bm-content-base -horizontal -centered -icon-right"
  >
    <div
      class="bm-c-alert__inner"
    >
      <div
        class="bm-c-alert__icon"
      >
        <div>
          icon
        </div>
      </div>
      <div
        class="bm-e-alert-content bm-text-base bm-c-alert__detail"
      >
        <h2>
          <strong>
            title
          </strong>
        </h2>
        <div
          class="bm-e-alert-content__spacer"
        />
        <div
          class="bm-e-alert-content__footer"
        >
          <div>
            footer
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Alert Component Snapshot Propsあり_レイアウト縦_中央揃えあり 1`] = `
<DocumentFragment>
  <div
    class="bm-c-alert bm-base bm-content-base -vertical -centered -icon-right"
  >
    <div
      class="bm-c-alert__inner"
    >
      <div
        class="bm-c-alert__icon"
      >
        <div>
          icon
        </div>
      </div>
      <div
        class="bm-e-alert-content bm-text-base bm-c-alert__detail -centered"
      >
        <h2>
          <strong>
            title
          </strong>
        </h2>
        <div
          class="bm-e-alert-content__spacer"
        />
        <div
          class="bm-e-alert-content__footer"
        >
          <div>
            footer
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Alert Component Snapshot Propsあり_レイアウト縦_中央揃えなし 1`] = `
<DocumentFragment>
  <div
    class="bm-c-alert bm-base bm-content-base -vertical -icon-right"
  >
    <div
      class="bm-c-alert__inner"
    >
      <div
        class="bm-c-alert__icon"
      >
        <div>
          icon
        </div>
      </div>
      <div
        class="bm-e-alert-content bm-text-base bm-c-alert__detail"
      >
        <h2>
          <strong>
            title
          </strong>
        </h2>
        <div
          class="bm-e-alert-content__spacer"
        />
        <div
          class="bm-e-alert-content__footer"
        >
          <div>
            footer
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Alert Component Snapshot Propsなし 1`] = `
<DocumentFragment>
  <div
    class="bm-c-alert bm-base bm-content-base -horizontal -icon-left"
  >
    <div
      class="bm-c-alert__inner"
    >
      <div
        class="bm-e-alert-content bm-text-base bm-c-alert__detail"
      />
    </div>
  </div>
</DocumentFragment>
`;
